keps/sig-apps/4368-support-managed-by-for-batch-jobs/README.md

```diff
@@ -362,6 +363,52 @@ We may come up with more validation rules during the implementation phase.
 The API comments to the Job status API fields will be updated to make the contract
 clear.
 
+We also plan to add a validation rule to make sure that the terminal conditions
+(`Failed` and `Complete`) are only added to Job when all pods are terminated.
+For that we plan to follow the approach described [below](#terminating-pods-and-terminal-job-conditions).
+
+#### Terminating pods and terminal Job conditions
+
+During the development process of Alpha in 1.30 we considered adding a
+validation rule enforcing that the Job terminal conditions (`Failed` or `Complete`)
+are only added when all pods are terminated (`status.terminating=0` and
+`status.ready=0`). However, the rule turned out to be violated by the built-in
+Job controller (see issue [#123775](https://github.com/kubernetes/kubernetes/issues/123775)).
+
+We are going to solve this issue by delaying the addition of the Job terminal
+conditions (Failed or Complete) until the pods are terminated
+(`terminating=0` and `ready=0`).
+
+One complication of this approach is that pod termination may take an
+arbitrarily long if the non-standard pod graceful termination period
+(`terminationGracePeriodSeconds`) is configured (30s by default). In order to
+give the API clients flexibility to know the fate of a Job as soon as possible
+we extend the scope for the following Job conditions:
+- `FailureTarget` - introduced in [Pod Failure Policy KEP](https://github.com/kubernetes/enhancements/issues/3329)
+- `SuccessCriteriaMet` - introduced in [Success Policy KEP](https://github.com/kubernetes/enhancements/issues/3998)
+
+Note that, with the new conditions an API client can now the fate of a Job
+earlier than currently, since adding the conditions will not wait for emptying
+of the `uncountedTerminatedPods` structure, as shown in the
+[experiment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2114710754).
+
+Since the fix is needed for this KEP and the
+[Pod Replacement Policy KEP](https://github.com/kubernetes/enhancements/issues/3939),
+we intend to protect the fix with the OR of the `JobManagedBy` and the
+`JobReplacementPolicy` feature gates.
+
+Note also that the fix impacts `CronJob` which in the `Forbid` concurrency
+policy (see [here](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#concurrency-policy)).
+However, it can be argued that the current behavior of the CronJob / Job integration
+does not match the expectations of the `Forbid` policy
+(see [comment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2115200217)),
+and can be handled as a bug.
```

Comment:
I add this note tentatively to non-goals.

I would be ok to support the new concurrency policy `ForbidActive` (or `SoftForbid`), but it isn't clear if needed, since the proposed changes better much the current documentation for the `Forbid` mode. OTOH, it might be a useful bailout config option in case there is a user who want to have the semantic as in your point 1.

cc @deads2k @soltysh @alculquicondor in case you have some additional opinion.

Reply:
I remember this from the success policy KEP.  As I recall, we chose to make the SuccessPolicy indicate `SuccessCriteriaMet` when we met criteria, but wait until all pods were terminated before transitioning to `Complete`.  Since the CronJob waits for complete, this makes the cronjob controller wait. Additionally, this is the safe default.

I'm ok with the idea of an addition concurrency policy, but I see it as separate from this KEP.

Reply:
> Maybe it would be good to also mention the pods with long termination times to the Risks section?

Done. PTAL.